editor/editor_data.cpp

```diff
@@ -509,6 +511,107 @@ EditorPlugin *EditorData::get_extension_editor_plugin(const StringName &p_class_
 	return plugin == nullptr ? nullptr : *plugin;
 }
 
+void EditorData::add_context_menu_plugin(ContextMenuSlot slot, const Ref<EditorContextMenuPlugin> &p_plugin) {
+	ContextMenu cm;
+	cm.p_slot = slot;
+	cm.plugin = p_plugin;
+	p_plugin->start_idx = context_menu_plugins.size() * EditorContextMenuPlugin::MAX_ITEMS;
+	context_menu_plugins.push_back(cm);
+}
+
+void EditorData::remove_context_menu_plugin(ContextMenuSlot slot, const Ref<EditorContextMenuPlugin> &p_plugin) {
+	for (int i = context_menu_plugins.size() - 1; i > -1; i--) {
+		if (context_menu_plugins[i].p_slot == slot && context_menu_plugins[i].plugin == p_plugin) {
+			context_menu_plugins.remove_at(i);
+		}
+	}
+}
+
+void EditorData::add_options_from_plugins(PopupMenu *p_popup, ContextMenuSlot p_slot, const Vector<String> &p_paths) {
+	bool add_separator = false;
+
+	for (int i = 0; i < context_menu_plugins.size(); i++) {
+		ContextMenu &cm = context_menu_plugins.get(i);
+		if (cm.p_slot != p_slot) {
+			continue;
+		}
+		if (!add_separator) {
+			add_separator = true;
+			p_popup->add_separator();
+		}
+
+		bool condition = cm.plugin->add_options(p_paths);
+		if (!condition) {
+			continue;
+		}
+		HashMap<String, EditorContextMenuPlugin::ContextMenuItem> &items = cm.plugin->context_menu_items;
+		int j = 0;
+		for (KeyValue<String, EditorContextMenuPlugin::ContextMenuItem> &E : items) {
+			EditorContextMenuPlugin::ContextMenuItem &item = E.value;
+			if (item.icon.is_valid()) {
+				p_popup->add_icon_item(item.icon, item.item_name, item.idx);
+			} else {
+				p_popup->add_item(item.item_name, item.idx);
+			}
+			if (item.shortcut.is_valid()) {
+				p_popup->add_shortcut(item.shortcut, item.idx);
+			}
+			j++;
+		}
+	}
+}
+
+template <typename T>
+void EditorData::invoke_plugin_callback(ContextMenuSlot p_slot, int p_option, const T &p_arg) {
+	Variant arg = p_arg;
+	Variant *argptr = &arg;
+
+	for (int i = 0; i < context_menu_plugins.size(); i++) {
+		if (context_menu_plugins[i].p_slot != p_slot || context_menu_plugins[i].plugin.is_null()) {
+			continue;
+			;
+		}
+		Ref<EditorContextMenuPlugin> plugin = context_menu_plugins[i].plugin;
+
+		HashMap<String, EditorContextMenuPlugin::ContextMenuItem> &items = plugin->context_menu_items;
+		for (KeyValue<String, EditorContextMenuPlugin::ContextMenuItem> &E : items) {
+			EditorContextMenuPlugin::ContextMenuItem &item = E.value;
+			if (p_option != item.idx || item.callable.is_null()) {
```

Comment:
```suggestion
			if (p_option != item.idx || !item.callable.is_valid()) {
```
Callable validity should check `is_valid` not `is_null` as they are not complementary